amazon step3,4 Carousel 구현

cocoa-salon opened the pull request:
- node.js를 이용해 로컬서버를 구축, ajax 요청에 따라 데이터를 수신하도록 구현하였습니다.(amazon_ajax_test.js)
- JS-HTML Templating 활용 시 DOMContentLoaded 이벤트에 등록된 리스너가 제대로 동작하지 않는 것 같아 load 이벤트를 이용하였습니다. 가급적 DOM 구성이 완료되는 시점에 온전한 스크립트 실행이 가능하려면 어떤 부분을 숙지해야 하는지요. 

감사합니다.

Commit:
1.carousel 기능 구현을 위한 html section-D 영역 추가

Commit:
2.carousel module을 위한 html 영역 추가 및 css 파일 별도 추가

3.Ajax 처리를 위한 더미 데이터 추가(JSON)

Commit:
4.carousel 애니메이션 처리를 위한 carousel module js 파일 별도 추가

5.Ajax 처리를 위한 로컬 서버 구동에 필요한 node.js 파일 추가

Commit:
6.PlansLayer, MorePlansLayer 클래스의 인스턴스 변수를 클래스 내부에 선언

7.PlansLayer, MorePlansLayer 클래스 내부 메소드 리팩토링

8.plans layer css 파일 별도 분리

9.plans layer css 클래스명 단축

crongro commented on carousel_module.js:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    prepareSlidingEvent() {
+        window.addEventListener("load", this.organizeCarouselLis.bind(this));
```
DOMContentloaded 이벤트로도 동작됐을텐데요.

원인이 이것인지 고민해보세요.
Ajax응답처리가 끝나기 전에 먼저 DOM을 접근한 것은 아닌가?

crongro commented on carousel_module.js:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    organizeCarouselLis() {
+        this.carouselPanels = document.querySelectorAll(".carousel-panels");
```
반복적으로 호출되는 것이 아니라면, 한줄짜리 메서드는 지양하는 것도 좋을 듯.

crongro commented on carousel_module.js:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    arrangeCarouselLis() {
+        for (let elemnt = 0; elemnt < this.carouselPanels.length; elemnt++) {
```
배웠죠.
nodelist도 for-of 를 쓸 수 있을거에요.

crongro commented on carousel_module.js:
```diff
@@ -0,0 +1,119 @@
+
+    // 자동 슬라이딩 시작
+    startAutoSlide() {
+        this.startTimerId = setInterval(this.setNextBtn.bind(this), 3000);
```
3000 이러한 매직넘버를 클래스에서 제거해보실래요? 
그러면 더욱 재사용가능한 클래스가 될 수 있습니다.

crongro commented on carousel_module.js:
```diff
@@ -0,0 +1,119 @@
+        this.carouselPanels.forEach(elemnt => {
+            elemnt.style.transform = `translateX(${this.distance}px)`
+        })
+        this.carouselUnit.firstElementChild.style.left = `${Number(`${lastItemLeft}`.match(/.\d+/)) + this.imageWidth}px`;
```
firstElementChild와 firstChild의 차이를 알고 계시죠?

cocoa-salon replied:
네, 부모 엘리먼트의 첫 번째 자식으로 텍스트 노드도 포함하느냐에 따라 차이가 있습니다. 이 점을 고려하여 firstElementChild 속성을 사용하였습니다.

crongro commented on carousel_module.js:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    setClickBtn(evt) {
+        this.eventIdentifier = evt.target.closest("div").classList[0];
```
classList[0] 요건 classname의 순서가 바뀌면 버그가 될 코드라 조금 위험합니다.

crongro commented on carousel_module.js:
```diff
@@ -0,0 +1,119 @@
+//amazon step4 carousel module js 
+
+// ajax 요청을 통한 JS 템플릿 완성 및 이미지 로드
+let ajaxElementTemplate = document.querySelector(".ajax-element-template").innerHTML;
```
ajax코드가 덩그러니 전역공간에 놓여있는데요.

Carousel관련된 코드라면 Carousel안으로 옮기는 게 어때요? 
constructor() {
  this.sendAjax();
}


그렇지 않다면 아래처럼 별도의 즉시실행함수로 위치할수도 있을거고요. (전역변수 없애기위해)
(function(){
 //ajax..code
})();

crongro commented on carousel_module.js:
```diff
@@ -0,0 +1,119 @@
+    // 자동 슬라이딩 재개
+    resumeAutoSlide(evt) {
+        if (this.eventIdentifier === "next-btn" || this.eventIdentifier === "prev-btn") {
+            this.resumeTimerId = setTimeout(this.startAutoSlide.bind(this), 2000);
```
여기 2000시간도 class밖으로 빼 보세요

crongro reviewed (changes requested):
method크기나 역할이 매우 좋아요. 잘 구현하셨고요.

코드에서 조금 수정할 부분이 보여서 리뷰남깁니다.
수정하시고 다시 push해주세요

Commit:
10.Ajax 코드 클래스 내부로 이동

-전역 공간에 위치한 코드 최소화

Commit:
11.window.onload 이벤트를 document.DOMContentLoaded 이벤트로 변경

-일부 변수 이름 변경(carouselPanelWidth) 및 초기화
-패널의 width 구하는 메소드 삭제

Commit:
12.기타 리팩토링

-시간설정 값(매직넘버) 인스턴스 객체로 초기화(slideTime)
-NodeList를 활용하는 메소드 for - of 문 변경(arrangeCarouselLis)

Commit:
13.기타 리팩토링

-setClickBtn 메소드에 사용하는 클래스에 대해 인덱스가 아닌 클래스 명으로 접근하도록 수정(3항 연산자 활용)
-시간설정 값(매직넘버)을 클래스 외부로 분리, 클래스 내에서 setSlideTime 메소드 호출을 통해 초기화

cocoa-salon commented:
보내주신 피드백에 따라 수정하여 푸시하였습니다. 확인 부탁드립니다. 

감사합니다.

Commit:
14.기타 리팩토링

-불필요한 DOMContentLoaded 이벤트 리스너 삭제, 단순 메소드 호출로 변경
-기존 HTML-JS template을 템플릿 리터럴 방식으로 변경

cocoa-salon commented:
오전 코드 리뷰 피드백을 반영하여 수정하였습니다. 

감사합니다.

crongro commented on carousel_module.js:
```diff
@@ -1,112 +1,122 @@
-
+// 슬라이드 시간 설정
+const slideTime = {
+    intervalTime: 3000,
```
const SLIDE_TIME = }
 ITERVAL :3000, RESTART : 2000
}

이렇게 대문자로 상수를 표시하고 중복된 문자를 줄이는 게 더 좋을 듯.

crongro commented on carousel_module.js:
```diff
@@ -1,112 +1,122 @@
+    constructor(slideTime) {
+        this.setSlideTime(slideTime);
+        this.callAjax();
+        this.carouselPanelWidth = 180;
```
이 숫자도 time처럼 인자로 받을 수 있겠죠.

crongro commented on carousel_module.js:
```diff
@@ -1,112 +1,122 @@
+                carouselModule.prepareSlidingEvent.call(carouselModule);
+            }
+        });
+        oReq.open("GET", "http://localhost:3000/img_src/ajax_imgs/imgMetadata");
```
URL을 함수안에 하드코딩하지 않고 인자로 받는게 좋겠죠.
개발단계에서의 URL과 실제서비스되는 서버의 URL 다를 수 있거든요. 
그래서 나중에 이부분을 수정해야하는데 함수안에 있으면  함수를 수정해야해서 좋은 방법은 아닙니다. 

저 위에 time처럼 분리해서 관리하고, 그곳을 수정하는게 안전해요.

crongro commented on carousel_module.js:
```diff
@@ -1,112 +1,122 @@
+    }
+
+    prepareSlidingEvent() {
+        document.addEventListener("DOMContentLoaded", this.arrangeCarouselLis());
```
이부분은 어제 리뷰드린대로 수정~

crongro reviewed (approved):
잘 수정했네요.
es6 classes는 지금 잘 사용하고 있으시고요. 
비동기 로직에 대해서도 잘 이해하고 계시도록 하세요.

여기서 머지할테니  지금 리뷰내용 참고 & 보완해두세요.

다음단계 진행하시면 될거 같습니다.
